stackdriver-core/lib/stackdriver/core/async_actor.rb

```diff
@@ -0,0 +1,347 @@
+        deadline = timeout ? ::Time.new.to_f + timeout : nil
+        synchronize do
+          until async_state == :stopped
+            cur_time = ::Time.new.to_f
+            return false if deadline && cur_time >= deadline
+            max_interval = @cleanup_options[:wait_interval]
+            interval = deadline ? deadline - cur_time : max_interval
+            interval = max_interval if interval > max_interval
+            @lock_cond.wait interval
+          end
+        end
+        true
+      end
+
+      ##
+      # @private Constructor to initialize MonitorMixin
+      #
+      def initialize
+        super()
+        @startup_lock = Mutex.new
+        @cleanup_options = {
+          wait_interval: WAIT_INTERVAL,
+          timeout: CLEANUP_TIMEOUT,
+          force: true
+        }
+      end
+
+      ##
+      # @private Wrapper method for running the async job. It requires classes
+      #   that include AsyncActor module to define a run_backgrounder method.
+      #   Then it runs a loop that checks for the state is workable (:running
+      #   or :suspended), which calls the run_backgrounder method. It ensures
+      #   the state variable gets set to :stopped when this method exists.
+      def async_run_job
+        fail "run_backgrounder method not defined" unless
+          respond_to? :run_backgrounder
+        until async_stopped?
+          run_backgrounder
+
+          return if async_stopping? && backgrounder_stoppable?
+        end
+      ensure
+        @async_state = :stopped
+        async_state_change
+      end
+
+      ##
+      # @private Ensures the child thread is started and kick off the job
+      #   to run async_run_job. Also make calls register_for_cleanup on the
+      #   async job to make sure it exits properly.
+      def ensure_thread
+        fail "async_actor not initialized" if @startup_lock.nil?
+        @startup_lock.synchronize do
+          if (@thread.nil? || !@thread.alive?) && @async_state != :stopped
+            @lock_cond = new_cond
+            AsyncActor.register_for_cleanup self
+            @thread = Thread.new do
+              async_run_job
+              AsyncActor.unregister_for_cleanup self
+            end
+            @async_state = :running
+            async_state_change
+          end
+        end
+      end
+
+      ##
+      # @private Set cleanup options.
+      #
+      # @param [Hash] *kwargs Hash of cleanup options. :timeout is the cleanup
+      #   wait timeout. :wait_interval is the cleanup wait interval. :force for
+      #   forcefully terminate actor when all other options fail.
+      def set_cleanup_options **kwargs
+        @cleanup_options.merge! kwargs
+      end
+
+      ##
+      # @private Default backgrounder stop condition when asked to be stopped
+      # gracefully. Called from #async_run_job when async actor state changes
+      # to :stopping
+      def backgrounder_stoppable?
+        true
+      end
+
+      def async_state_change
```

Comment:
Unlike `backgrounder_stoppable?`, this method can be potentially extendable to handle other things when the actor state changes. Maybe if I move the `@lock_cond.broadcast calls` into this function, then it would make more sense. What do you think?